### PR TITLE
Streams: tests for teeing readable byte streams

### PR DESCRIPTION
**Pull request by MattiasBuelens**
See whatwg/streams#1114 for the accompanying spec change.

### CODE REVIEW

File: streams/readable-byte-streams/tee.any.js [hunk 1]
```diff
@@ -0,0 +1,776 @@
...
+  {
+    const result = await reader1.read(new Uint8Array(1));
+    assert_equals(result.done, false, 'done');
+
+    const view = result.value;
+    assert_equals(view.constructor, Uint8Array, 'value.constructor');
+    assert_equals(view.buffer.byteLength, 1, 'value.buffer.byteLength');
+    assert_equals(view.byteOffset, 0, 'value.byteOffset');
+    assert_equals(view.byteLength, 1, 'value.byteLength');
+    assert_equals(view[0], 0x01);
```
Comment (domenic) [inline streams/readable-byte-streams/tee.any.js:45]: Maybe abstract this into a helper?

File: streams/readable-byte-streams/tee.any.js [hunk 2]
```diff
@@ -0,0 +1,776 @@
...
+  assert_equals(view1.constructor, Uint8Array, 'reader1 value.constructor');
+  assert_equals(view2.constructor, Uint8Array, 'reader2 value.constructor');
+  assert_equals(view1.buffer.byteLength, 1, 'reader1 value.buffer.byteLength');
+  assert_equals(view1.buffer.byteLength, 1, 'reader2 value.buffer.byteLength');
+  assert_array_equals([...view1], [0x01], `reader1 value`);
+  assert_array_equals([...view2], [0x01], `reader2 value`);
+
+  assert_not_equals(view1.buffer, view2.buffer, 'chunks should have different buffers');
+
+}, 'ReadableStream teeing with byte source: chunks should be cloned for each branch');
```
Comment (domenic) [inline streams/readable-byte-streams/tee.any.js:119]: The spec only clones for branch 2, right? But branch 1 will get transferred. So it's worth checking that both branches' chunks are !== the original enqueued chunk.

File: streams/readable-byte-streams/tee.any.js [hunk 3]
```diff
@@ -0,0 +1,776 @@
...
+    }
+  });
+
+  const [branch1, branch2] = rs.tee();
+  await Promise.all([
+    promise_rejects_exactly(t, theError, branch1.cancel()),
+    promise_rejects_exactly(t, theError, branch2.cancel())
+  ]);
+
+}, 'ReadableStream teeing with byte source: failing to cancel the original stream should cause cancel() to reject on branches');
```
Comment (domenic) [inline streams/readable-byte-streams/tee.any.js:316]: For these do you think it'd be easier to parameterize over bytes/non-bytes in a single file?

**Review by domenic — APPROVED** [review]
Generally looks good and quite exhaustive; impressive as always.

Were you able to check coverage on the reference implementation?

File: streams/readable-byte-streams/tee.any.js (hunk 3 above)
Comment (MattiasBuelens) [reply streams/readable-byte-streams/tee.any.js:316]: The order in which I wrote these non-bytes/bytes tests does not match the order in which they landed in the PRs. 😛

You're right, I'll see if I can make these tests parameterized. 🙂

**Comment by MattiasBuelens** [comment]
> Were you able to check coverage on the reference implementation?

Riiiight, we have a script for that! 😅 I haven't checked it yet, I'll have a look.

**Comment by MattiasBuelens** [comment]
Looks like `npm run coverage` is broken in the reference implementation. 😬

Istanbul works by [hooking into `require()`](https://github.com/istanbuljs/nyc/blob/v15.1.0/README.md#selecting-files-for-coverage). Previously, we'd just `require()` the reference implementation and [expose that to jsdom's `window`](https://github.com/whatwg/streams/blob/99ffba7a632c261566c12da60bb35800a42e3b3f/reference-implementation/run-web-platform-tests.js#L59-L63) in the test runner. But nowadays we [bundle the whole thing with Browserify first](https://github.com/whatwg/streams/blob/9beefc2ac041a1a848088962d56eb03e89a8bf07/reference-implementation/run-web-platform-tests.js#L87-L91) and then [eval the bundle](https://github.com/whatwg/streams/blob/9beefc2ac041a1a848088962d56eb03e89a8bf07/reference-implementation/run-web-platform-tests.js#L61) to get it in jsdom's `window`. Since we're no longer using `require()`, Istanbul doesn't process any of our code and can't report any coverage.

Not sure if/how we want to fix this? 😕

**Comment by MattiasBuelens** [comment]
Right, managed to fix the coverage script. Some cases aren't tested yet:
* Pull with a BYOB reader, then pull with a default reader (should swap the internal reader)
* Pull with a default reader, then close the stream while the other branch has a pending BYOB read (should call `respond(0)`)
* Pull with a default reader, then error the stream
* Pull with a BYOB reader, then close the stream while the other branch has a pending BYOB read (should call `respond(0)`)
* Pull multiple times with one reader (should hit the `if (reading === true)` check in the pull algorithms)

**Review by domenic — APPROVED** [review]
LGTM. Not sure why the bots are failing; try rebasing?

@ricea feel free to take a look

**Pull request merged**